all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+  fn max(&self) -> Option<usize>;
+}
+
+/// The user would be allowed to provide either just a slot supplier (using the standard fixed-size cache),
+/// or one of these to handle both themselves.
+/// 
+/// I'm a bit at a loss for a good name
+trait WorkflowSlotAndCacheManager: WorkflowTaskSlotSupplier + WorkflowCacheSizer {}
+```
+
+## Flow
+
+Some context on the current flow (at least in Core). The cache only applies to workflow tasks. This flow should remain
+the same after these changes. The flow makes it clear that a misbehaving `WorkflowSlotAndCacheManager` which allows lots
+of slots but keeps saying the cache should not expand could result in buffering tasks and generally gumming things
+up.
+
+```mermaid
+sequenceDiagram
+  participant T as Temporal
+  participant W as Worker
+  participant S as Slots
+  participant C as CacheSizer
+  W->>S: reserve_slot
+  S-->>W: Allowed
+  W->>T: Poll
+  T-->>W: Response
+  
+  alt Workflow is not already in cache
+    W->>C: `can_allow_workflow` ?
+    C-->>W: I'm full
+    W->>W: Buffer task
+    W->>W: Wait for some workflow to be idle (maybe there is already)
+    W->>C: `eviction_hint`
+    C-->>W: Evict This guy
+    W->>W: Evict workflow
+    W->>C: `evicted_workflow`
+    W->>W: Cache new workflow
+  end
+  
+  W->>S: `mark_slot_used`
+  W->>W: Process workflow task
+  W->>S: `release_slot`
+```
+
+## Default implementations
+
+We should provide a few default implementations:
+
+* Semaphore based SlotSupplier: This is the current implementation. There's a fixed number of slots, that's it.
+* Resource based SlotSupplier: This implementation can look at how close the worker is to some resource limit
+  (% of CPU and/or memory used), and hand out slots if it's still below that limit. This only makes sense for
+  activity slot suppliers, since workflows use little CPU, and the memory they use is the concern of cache
+  management, not the slot supplier.
+* Memory based WorkflowCacheSizer: Like above, but for the sizing of the workflow cache.
+
+## Metrics
+
+Users can add their own metrics behind their implementations now, which is great - but we can also provide some out of
+the box. We can always provide `slots_in_use` by type. Right now we emit the inverse of this,
+`worker_task_slots_available` - we can keep emitting that with a bundled implementation that replicates the old
+fixed-number based implementation. In fact, any implementation which returns a value from `available_slots` we can make
+emit the metric automatically.
+
+We can also of course keep emitting the current number of cached workflows.
+
+
+## Where it hooks up and the language barrier
+
+Staying in Rust for a moment, the slot supplier implementations can be hooked up on a per-worker basis through
+`WorkerConfig`, as expected. We would add:
+
+```rust
+pub struct WorkerConfig {
+  // ... existing fields
+  #[builder(default)]
+  pub workflow_task_slot_and_cache: Option<Arc<dyn WorkflowSlotAndCacheManager>>,
```

Comment:
I think slot supplier and cache should be provided as separate option, even if the implementer chooses to have the same struct impl both

Reply:
I'm OK with that but curious to hear why you think so.

I made it this way because I liked the idea of the user explicitly combining the workflow things into one piece and providing that, since it makes it clear they are somewhat related to each other, but it's not the most important reason.

Reply:
This may be language specific, but many may not care about cache size in their impl.

I'm ok keeping them together, but it becomes harder to justify the separate interfaces then if they must always both be implemented anyways. So now I'm wondering if we want to change the naming a bit from `SlotSupplier` to a more general purpose `Limiter`, `Tuner`, `Controller`, `TaskManager`, etc. So:

* `ActivitySlotSupplier` is instead `ActivityController` and has the slot supplier stuff
* `WorkflowSlotSupplierAndCacheManager` is instead `WorkflowController` and has slot supplier stuff and cache manager stuff

Basically this lifts the low-level notion of "slot supplying" and "cache managing" to a general purpose controller for limiting/controlling work. I think it will also be more clear to users to `controller = new ResourceBasedWorkerController()` + `myWorkerConfig.WorkflowController = controller`. They don't have to care what slots are or that the interface is limited to slot supplying only or whatever.

Reply:
Not a fan of combining slot and cache management. Slots are per worker, where the cache is per process in Go and per worker factory in Java. I would also like to leave the door open to how we map the cache to workers in the SDK ideally so we can have per worker cache in Go if desired.